quickjs.c

```diff
@@ -45531,180 +45531,180 @@ static int js_json_to_str(JSContext *ctx, JSONStringifyContext *jsc,
                 goto exception;
         } else {
             sep = js_dup(jsc->empty);
             sep1 = js_dup(jsc->empty);
         }
-        v = js_array_push(ctx, jsc->stack, 1, &val, 0);
-        if (check_exception_free(ctx, v))
-            goto exception;
+
         ret = JS_IsArray(ctx, val);
         if (ret < 0)
             goto exception;
+            
         if (ret) {
+            int64_t len;
             if (js_get_length64(ctx, &len, val))
                 goto exception;
-            string_buffer_putc8(jsc->b, '[');
-            for(i = 0; i < len; i++) {
-                if (i > 0)
-                    string_buffer_putc8(jsc->b, ',');
-                string_buffer_concat_value(jsc->b, sep);
-                v = JS_GetPropertyInt64(ctx, val, i);
-                if (JS_IsException(v))
-                    goto exception;
-                /* XXX: could do this string conversion only when needed */
-                prop = JS_ToStringFree(ctx, js_int64(i));
-                if (JS_IsException(prop))
+                
+            if (len < 0 || len > INT32_MAX) {
+                JS_ThrowRangeError(ctx, "array length out of range");
+                goto exception;
+            }
+
+            if (string_buffer_putc8(jsc->b, '[') < 0)
+                goto exception;
+
+            for(int64_t i = 0; i < len; i++) {
+                if (i > 0) {
+                    if (string_buffer_putc8(jsc->b, ',') < 0)
+                        goto exception;
+                }
+                
+                JSValue elem = JS_GetPropertyInt64(ctx, val, i);
+                if (JS_IsException(elem))
                     goto exception;
-                v = js_json_check(ctx, jsc, val, v, prop);
-                JS_FreeValue(ctx, prop);
-                prop = JS_UNDEFINED;
-                if (JS_IsException(v))
+                    
+                elem = js_json_check(ctx, jsc, val, elem, JS_NewInt64(ctx, i));
+                if (JS_IsException(elem))
                     goto exception;
-                if (JS_IsUndefined(v))
-                    v = JS_NULL;
-                if (js_json_to_str(ctx, jsc, val, v, indent1))
+
+                if (js_json_to_str(ctx, jsc, val, elem, indent1) < 0)
                     goto exception;
             }
-            if (len > 0 && !JS_IsEmptyString(jsc->gap)) {
-                string_buffer_putc8(jsc->b, '\n');
-                string_buffer_concat_value(jsc->b, indent);
-            }
-            string_buffer_putc8(jsc->b, ']');
+
+            if (string_buffer_putc8(jsc->b, ']') < 0)
+                goto exception;
         } else {
-            if (!JS_IsUndefined(jsc->property_list))
-                tab = js_dup(jsc->property_list);
-            else
-                tab = js_object_keys(ctx, JS_UNDEFINED, 1, &val, JS_ITERATOR_KIND_KEY);
-            if (JS_IsException(tab))
+            JSPropertyEnum *props = NULL;
+            uint32_t prop_count = 0;
+
```

Comment:
This array should be defined at the outer scope so the objects can be freed in case of `exception`.